apps/gallery/js/MetadataParser.js

```diff
@@ -303,7 +303,16 @@ var metadataParser = (function() {
         // thumbnail and an external preview image. Note that progressive
         // jpegs will have a type 'pjpeg' and will be handled below
         // like png images.
-        if (metadata.type === 'jpeg') {
+
+        // For jpeg images with sizes more than 2.5 mb, we need
+        // to create preview on sdcard to avoid flash while opening image.
+        // See bug 1181290.
+        // This constraint can be removed when gecko supports downscale during
+        // decode by canvas size for offscreen images. This will
+        // facilitate removing #moz-sample-size and using ddd.
+        // See bug 1206206.
+
+        if (metadata.type === 'jpeg' && file.size < 2.5 * 1024 * 1024) {
```

Comment:
Please use the image size, not the file size here. I'd suggest 5 or 8 megapixels as the limit, not 2.5 megabytes:
```
if (metadata.type === 'jpeg' && metadata.width * metadata.height < 5 * 1024 * 1024) {
```